lib/ui/lerp.dart

```diff
@@ -48,3 +48,18 @@ int _clampInt(int value, int min, int max) {
     return value;
   }
 }
+
+/// Same as [num.clamp] but optimized for non-null [double].
+double clampDouble(double x, double min, double max) {
```

Comment:
I don't think we should add this to the engine, given that we don't need to use it here. 

dart:ui is very difficult to remove code from. Whereas In flutter land we could concievably switch back to double.clamp if the VM team can fix the implementation.

Reply:
Looks like there is some places in the engine I can use this too.  It's right next to other clamp implementations so it makes sense to me.  I planned on writing up some tests before putting it into review, I'll move over the engine to use this too then we can see how it looks.

Reply:
Those are private though.

Reply:
`lerpDouble` isn't.  That seems like a pretty clear precedent for `clampDouble` which I tried to follow.  It's used in the engine and in the framework.

Reply:
Yea turns out there was only one place where `clampDouble` could currently be used in the engine.  It isn't hard to imagine it wanting to use it more in the future though.

Reply:
If anything the desire is to have `lerpDouble` moved _out_ of the engine into the SDK.

I think it'd be better to have these be private. Ideally at some point Dart fixes `double.clamp` so it's just as fast, and we don't want to have to deprecate/remove it (like with hashValues/hashList).

Reply:
(also, private functions in dart:ui can be used throughout dart:ui, even if not in the same file)

Reply:
I would put it in foundation, or since the big impact comes from usage in layout you could make it a private function in object.dart or proxy_box.dart

Reply:
+1

Reply:
Okay, I can look into putting it back in there later today.  I think I ran into some problems where certain code didn't seem to have access to foundation, or I just didn't see an example of how to do it yet.  I don't want to make it private to object.dart because if you look through the usages of it there is a lot of potentially good stuff like animations where it can make a difference we weren't event measuring.